### ISSUE TITLE
训练过程中出现 paddle.fluid.core.EnforceNotMet: Enforce failed. Expected framework::product(new_dim) / new_dim[0] == framework::product(t->dims()) / t->dims()[0], but received framework::product(new_dim) / new_dim[0]:324 != framework::product(t->dims()) / t->dims()[0]:576.

### ISSUE THREAD

**Issue by renhui19931001** [PaddlePaddle/Paddle#18780]
本地训练resnet50的网络，在训练的时候每个batchsize的image size都不一致，训练过程中会随机位置出现错误，环境如下http://yq01-sys-hic-k40-box-0004.yq01.baidu.com:8388/v1/slurmjobs/39729/workspace

**Comment by renhui19931001** [comment]
训练环境paddle 1.3 gpu, k40, p40, v100训练能稳定复现问题
具体的错误代码是:
File "./run_train_scripts/local_train/same_4w5_delf.py", line 58, in <module>
trainmain()
File "./run_train_scripts/local_train/same_4w5_delf.py", line 55, in trainmain
train.main()
File "thirdparty/paddlemodels/deep_local/trainfast_aiflow_local.py", line 347, in main
train_async(args)
File "thirdparty/paddlemodels/deep_local/trainfast_aiflow_local.py", line 289, in train_async
lr, loss, acc1, acc5, score = train_exe.run(fetch_list=train_fetch_list)
File "/home/slurm/job/tmp/job-39729/paddle-release/python-gcc482-paddle/lib/python2.7/site-packages/paddle/fluid/parallel_executor.py", line 297, in run
self.executor.run(fetch_list, fetch_var_name)
paddle.fluid.core.EnforceNotMet: Enforce failed. Expected framework::product(new_dim) / new_dim[0] == framework::product(t->dims()) / t->dims()[0], but received framework::product(new_dim) / new_dim[0]:324 != framework::product(t->dims()) / t->dims()[0]:576.
at [/sharefolder/duyuning/Paddle/paddle/fluid/framework/lod_tensor.cc:394]
PaddlePaddle Call Stacks: 
0 0x7fe1a627c8cdp void paddle::platform::EnforceNotMet::Init<std::string>(std::string, char const*, int) + 365
1 0x7fe1a627d730p paddle::platform::EnforceNotMet::EnforceNotMet<char const*, char const*, char const*, char const*, std::string, char const*, std::string, std::string>(char const*, int, char const*, char const*, char const*, char const*, std::string, char const*, std::string, std::string) + 144
2 0x7fe1a7814387p paddle::framework::LoDTensor::MergeLoDTensor(std::vector<paddle::framework::LoDTensor const*, std::allocator<paddle::framework::LoDTensor const*> > const&, boost::variant<paddle::platform::CUDAPlace, paddle::platform::CPUPlace, paddle::platform::CUDAPinnedPlace, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_>) + 1335
3 0x7fe1a7918ab5p paddle::framework::details::FetchOpHandle::WaitAndMergeCPUTensors() const + 309
4 0x7fe1a7918f18p paddle::framework::details::FetchOpHandle::RunImpl() + 936
5 0x7fe1a7959786p paddle::framework::details::OpHandleBase::Run(bool) + 118
6 0x7fe1a78df78dp
7 0x7fe1a728fb43p std::_Function_handler<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> (), std::__future_base::_Task_setter<std::unique_ptr<std::__future_base::_Result<void>, std::__future_base::_Result_base::_Deleter>, void> >::_M_invoke(std::_Any_data const&) + 35
8 0x7fe1a6dc0ff7p std::__future_base::_State_base::_M_do_set(std::function<std::unique_ptr<std::__future_base::_Result_base, std::__future_base::_Result_base::_Deleter> ()>&, bool&) + 39
9 0x7fe2068df973p pthread_once + 83
10 0x7fe1a78de472p
11 0x7fe1a6dc2284p ThreadPool::ThreadPool(unsigned long)::{lambda()#1}::operator()() const + 404
12 0x7fe1f51fe8a0p
13 0x7fe2068da1c3p
14 0x7fe205f0212dp clone + 109

训练代码：
![image](https://user-images.githubusercontent.com/19266892/61776386-8040c300-ae2d-11e9-9b92-c394239c3c4b.png)
![image](https://user-images.githubusercontent.com/19266892/61776424-8f277580-ae2d-11e9-8fcc-fb47e0bc7462.png)

**Comment by renhui19931001** [comment]
reader:
        """used to train normal softmax """
        reader = create_rawdatareader(settings, which, "./cached/")
        loss_name = settings.loss_name
        batchsize = settings.train_batch_size
        showiter = settings.display_iter_step
        random_max_size = settings.random_max_size
        random_min_size = settings.random_min_size

        if which != "train":
            for imgvalue, label in reader():
                yield imgvalue, int(label)
        elif loss_name == "softmax" or loss_name == "arcmargin":

            cnt = 0.0
            while True:
                for imgvalue, label in reader():
                    if cnt % (showiter * batchsize) == 0:
                        crop_size = random.randint(random_min_size, random_max_size)
                        crop_size_norm = normwidth(crop_size, settings.imgfactor)
                        cnt = 0
                    cnt += 1
                    yield imgvalue, int(label), crop_size_norm

**Comment by renhui19931001** [comment]
reader:
        """used to train normal softmax """
        reader = create_rawdatareader(settings, which, "./cached/")
        loss_name = settings.loss_name
        batchsize = settings.train_batch_size
        showiter = settings.display_iter_step
        random_max_size = settings.random_max_size
        random_min_size = settings.random_min_size

        if which != "train":
            for imgvalue, label in reader():
                yield imgvalue, int(label)
        elif loss_name == "softmax" or loss_name == "arcmargin":

            cnt = 0.0
            while True:
                for imgvalue, label in reader():
                    if cnt % (showiter * batchsize) == 0:
                        crop_size = random.randint(random_min_size, random_max_size)
                        crop_size_norm = normwidth(crop_size, settings.imgfactor)
                        cnt = 0
                    cnt += 1
                    yield imgvalue, int(label), crop_size_norm

train:
train_reader = paddle.batch(reader.train(args), batch_size=train_batch_size, drop_last=True)
    train_py_reader.decorate_paddle_reader(train_reader)
    train_exe = fluid.ParallelExecutor(main_program=train_prog,
        use_cuda=args.use_gpu, loss_name=train_cost.name)
    totalruntime = 0
    train_py_reader.start()
    iter_no = 0
    train_info = [0, 0, 0, 0]
    while iter_no <= args.total_iter_num:
        t1 = time.time()
        lr, loss, acc1, acc5, score = train_exe.run(fetch_list=train_fetch_list)
        #print(score.shape)

**Comment by guoshengCS** [comment]
这个报错的原因是fetch的时候多卡的内容没办法merge起来，请问fetch的内容中有各个卡上除了batch_size以外的维度shape不一样的吗

**Comment by renhui19931001** [comment]
应该没有吧，主要是多个训练任务都出现这个问题，后来我把size改成固定224了，还是报这个问题，我贴上我后面的处理代码:
    def net(self, input):
        input.stop_gradient = True

        #add param in feature reduce
        if self.fea_reduce == 0:
            feature = self.conv_bn_layer(input, self.mid_len, 1, name="delf_conv1")
        elif self.fea_reduce == 1:
            # add multi down
            feature_st1 = conv_reduce(input, "delf_down_r1", num_filters = 512, filter_size=3, padding=1)
            feature = conv_reduce(feature_st1, "delf_down", num_filters = self.mid_len, filter_size=1, padding=0)

        #add param to determ where up sample feature map
        if self.upsample:
            feature = self._upsampling(feature, self.upsample)

        #detern where train float or hash descriptor
        if self.use_hash:
            fea_norm = fluid.layers.tanh(feature)
        else:
            fea_norm = fluid.layers.l2_normalize(feature, axis=1, name="delffea_norm")

        midtmp = fluid.layers.conv2d(
            input=feature, num_filters=self.mid_len // 2, filter_size=1, stride=1,
            padding=0, groups=1, act="relu", bias_attr=False, name="delf_conv2")
        score = fluid.layers.conv2d(
            input=midtmp, num_filters=1, filter_size=1, stride=1,
            padding=0, groups=1, act="relu", bias_attr=False, name="delf_conv3")

        output = fluid.layers.elementwise_mul(fea_norm, fluid.layers.expand(score, expand_times=[1, fea_norm.shape[1], 1, 1]))

        output = fluid.layers.pool2d(
            input=output, pool_size=14, pool_type='avg', global_pooling=True)

        return [output, fea_norm, score]

**Comment by renhui19931001** [comment]
单scale训练报错的日志为：http://yq01-sys-hic-k40-box-0004.yq01.baidu.com:8388/v1/slurmjobs/41217/workspace
http://yq01-sys-hic-k40-box-0022.yq01.baidu.com:8388/v1/slurmjobs/41212/workspace

paddle环境 gpu 1.3, 在p40, k40, v100上都随机出现，可能有时候训练完了都不挂，大部分会随机挂掉

**Comment by guoshengCS** [comment]
看了下给的链接，那个报错确实是因为fetch的时候多卡的内容没办法merge，如果出现的频率很高的话，可否试下run的时候不进行fetch看是否会有问题。
另外可否说下fetch的具体内容有哪些呢，都是什么shape的呢

**Comment by renhui19931001** [comment]
fetch的内容有train_fetch_list = [global_lr.name, train_cost.name, train_acc1.name, train_acc5.name, score.name]，最后一个是featuremap,在训练空间中./thirdparty/paddlemodels/deep_local/trainfast_aiflow_local.py

**Comment by renhui19931001** [comment]
是不是最后的featuremap score（1, 64, size, size）的内容在训练的过程中不能fetch出来呢

**Comment by guoshengCS** [comment]
如果每个batchsize的image size都不一致，因为每张卡上都对应了一个batch的数据（使用pyreader的话会是这种情况），那每张卡上的featuremap是否会大小不同呢，这样的话确实没办法fetch，如果要fetch这个`featuremap score（1, 64, size, size)`需要保证每张卡上的size都是相同的。可以试下先不fetch这个featuremap看下

**Comment by renhui19931001** [comment]
我实现的时候，reader里面，利用了keep_order=True，保证整体顺序，然后再每个batchsize的图片内size一致，但是每个batchsize训练不一致
def createreader(settings, which):
    def normal_reader():
        """used to train normal softmax """
        reader = create_rawdatareader(settings, which, "./cached/")
        loss_name = settings.loss_name
        batchsize = settings.train_batch_size
        showiter = settings.display_iter_step
        random_max_size = settings.random_max_size
        random_min_size = settings.random_min_size

        if which != "train":
            for imgvalue, label in reader():
                yield imgvalue, int(label)
        elif loss_name == "softmax" or loss_name == "arcmargin":

            cnt = 0.0
            while True:
                for imgvalue, label in reader():
                    if cnt % (showiter * batchsize) == 0:
                        crop_size = random.randint(random_min_size, random_max_size)
                        crop_size_norm = normwidth(crop_size, settings.imgfactor)
                        cnt = 0
                    cnt += 1
                    yield imgvalue, int(label), crop_size_norm

    image_mapper = functools.partial(process_image_delf,
            mode=which, color_jitter=False, rotate=True,
            minsize=settings.random_min_size, maxsize=settings.random_max_size)
    img_process_reader = paddle.reader.xmap_readers(
            image_mapper, normal_reader, 8, 1000, order=True)
    def to_paddle_reader():
        """to_paddle_reader"""
        for data in img_process_reader():
            if data is not None:
                yield data
    #return to_paddle_reader
    return img_process_reader

**Comment by guoshengCS** [comment]
如果你的reader产生的是一张卡上的数据，那八卡训练你会产生八次数据，要看下这八份数据的image size是否一样

**Comment by renhui19931001** [comment]
八分数据的image size是一样的，不然启动训练都起不了，我训练分类过程中，设置batchsize是256，平均8卡，每张卡上使用1,32,size,size的feature map，现在问题不在于起不起来，而是在于训练到某个epoch之后，随机可能会挂掉了。

**Comment by guoshengCS** [comment]
麻烦还是先试下不fetch featuremap看下吧，这个是目前定位的问题所在，先验证下是否是这个问题

**Comment by renhui19931001** [comment]
问题应该解决了，谢啦，起了6个任务5小时无core，应该就是在训练过程中，fetch部分的featuremap在paddle中不支持，不知道算不算bug。
